test/unit/math/rev/core/thread_stack_instance_test.cpp

```diff
@@ -0,0 +1,84 @@
+#include <gtest/gtest.h>
+#include <stan/math/rev/core.hpp>
+
+#include <thread>
+
+TEST(thread_stack_instance, initialize) {
+  using stan::math::ChainableStack;
+
+  // the main thread must be initialized by the time this code is
+  // reached. This will actually segfault if this is not the case.
+  // The pointer to the reference returned must evaluate to true.
+  EXPECT_TRUE(&ChainableStack::instance());
+
+  ChainableStack::AutodiffStackStorage& main_ad_stack
+      = ChainableStack::instance();
+
+#ifdef STAN_THREADS
+  auto thread_tester = [&]() -> void {
+    ChainableStack thread_instance;
+    EXPECT_TRUE(&ChainableStack::instance());
+    EXPECT_TRUE(&ChainableStack::instance() != &main_ad_stack);
+  };
+#else
```

Comment:
Could you narrow the scope of the `ifdef` here and in the other ones as well to highlight what is changing? It took me a few minutes to spot the difference here :P

Reply:
Will do. NP.